linkerd/app/outbound/src/lib.rs

```diff
@@ -129,49 +129,111 @@ impl<S> Outbound<S> {
         P::Error: Send,
         I: io::AsyncRead + io::AsyncWrite + io::PeerAddr + std::fmt::Debug + Send + Unpin + 'static,
     {
+        let tcp_endpoint = self.clone().push_tcp_endpoint().into_inner();
+        let http_endpoint = self.clone().push_tcp_endpoint().into_inner();
```

Comment:
the `http_endpoint` doesn't actually `push_http_endpoint`? Shouldn't it? If not, do we really need to pass two of these to `into_server_with`?

Why is this necessary? There's a _lot_ of type boilerplate for this split. Why?

Reply:
This is for tests — it's in order to allow us to configure the proxy with two separate endpoint stacks, one which is used for HTTP and another which is used for TCP forwards. This allows us to e.g. fail the HTTP tests if a connection is treated as raw TCP when it shouldn't be. 

Previously, the tests did this by just duplicating the code from `Outbound::into_server`. I wanted to change this so that there wasn't this duplication that has to be kept in sync — in particular, it was very easy to add new layers or change the structure of the stack built in `into_server`, but neglect to change the tests. Some of the test code had already silently become out of sync and wasn't testing a representative proxy, so rather than update the duplicated stacks, I thought it was better to just change `into_server` so that the tests could use it.